src/imex/transfer.rs

```diff
@@ -0,0 +1,539 @@
+    ///
+    /// This will acquire the global "ongoing process" mutex, which can be used to cancel
+    /// the process.
+    ///
+    /// [`Accounts::stop_io`]: crate::accounts::Accounts::stop_io
+    pub async fn prepare(context: &Context) -> Result<Self> {
+        e2ee::ensure_secret_key_exists(context)
+            .await
+            .context("Private key not available, aborting backup export")?;
+
+        // Acquire global "ongoing" mutex.
+        let cancel_token = context.alloc_ongoing().await?;
+        let context_dir = context
+            .get_blobdir()
+            .parent()
+            .ok_or(anyhow!("Context dir not found"))?;
+        let dbfile = context_dir.join(DBFILE_BACKUP_NAME);
+        if fs::metadata(&dbfile).await.is_ok() {
+            fs::remove_file(&dbfile).await?;
+            warn!(context, "Previous database export deleted");
+        }
+        let res = tokio::select! {
+            biased;
+            res = Self::prepare_inner(context, &dbfile) => {
+                match res {
+                    Ok(slf) => Ok(slf),
+                    Err(err) => {
+                        error!(context, "Failed to set up second device setup: {:#}", err);
+                        Err(err)
+                    },
+                }
+            },
+            _ = cancel_token.recv() => Err(format_err!("cancelled")),
+        };
+        let (provider, ticket) = match res {
+            Ok((provider, ticket)) => (provider, ticket),
+            Err(err) => {
+                context.free_ongoing().await;
+                return Err(err);
+            }
+        };
+        let handle = tokio::spawn(Self::watch_provider(
+            context.clone(),
+            provider,
+            cancel_token,
+            dbfile,
+        ));
+        let slf = Self { handle, ticket };
+        let qr = slf.qr();
+        *context.export_provider.lock().expect("poisoned lock") = Some(qr);
+        Ok(slf)
+    }
+
+    /// Creates the provider task.
+    ///
+    /// Having this as a function makes it easier to cancel it when needed.
+    async fn prepare_inner(context: &Context, dbfile: &Path) -> Result<(Provider, Ticket)> {
+        // Generate the token up front: we also use it to encrypt the database.
+        let token = AuthToken::generate();
+        context.emit_event(SendProgress::Started.into());
+        export_database(context, dbfile, token.to_string())
+            .await
+            .context("Database export failed")?;
+        context.emit_event(SendProgress::DatabaseExported.into());
+
+        // Now we can be sure IO is not running.
+        let mut files = vec![DataSource::with_name(
+            dbfile.to_owned(),
+            format!("db/{DBFILE_BACKUP_NAME}"),
+        )];
+        let blobdir = BlobDirContents::new(context).await?;
+        for blob in blobdir.iter() {
+            let path = blob.to_abs_path();
+            let name = format!("blob/P{}", blob.as_file_name());
```

Comment:
maybe add a comment, what the `blob/P` prefix means

Reply:
:scream: this is a bug and is not caught by the tests currently

Reply:
fixed now, with test